### PR TITLE
Creates an empty logger library

### PR DESCRIPTION
**Pull request by RekhaAparna01**
Creates an empty logger library 

This commit is to create a logger library
The logger library is used to log messages to buffer 
based on the severity level of the message.

Test:
Tested by adding debug traces in the library

Change-Id: I26288f8a00415b519aa9d8c0304e8f038e0febcd
Signed-off-by: RekhaAparna01 <vrekhaaparna@ibm.com>

### CODE REVIEW

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
There are two empty files. Please get rid of those if not required,

File: logger/src/logger.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+#include <iostream>
+#include <sstream>
+#include <string>
+
+namespace logging
+{
+
+void logMessage(std::string message, const std::source_location& location)
+{
+    std::ostringstream logMsg;
```
Comment (SunnySrivastava1984) [inline logger/src/logger.cpp:13]: This is a local variable so how the API is appending to the this?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: logger/src/logger.cpp (hunk 1 above)
Comment (RekhaAparna01) [reply logger/src/logger.cpp]: Done.. changed it to global variable

File: src/panel_app_main.cpp [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #include "bus_monitor.hpp"
 #include "button_handler.hpp"
 #include "const.hpp"
+#include "logger.hpp"
```
Comment (SunnySrivastava1984) [inline src/panel_app_main.cpp:5]: Why this header was added?

File: logger/include/logger.hpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+{
+    INFO,
+    CRITICAL
+};
+
+class Logger
+{
+  public:
+    Loglevel level;
+    std::stringstream logmessage;
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:18]: Please refer coding guidelines document.

File: logger/include/logger.hpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        }
+        return *this;
+    }
+    /**
+     * @brief Operator overloading function to append message to the buffer
+     *
+     * param[in]- Information to be appended to the buffer
+     */
+
+    template <class T>
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:48]: Why template is needed here?

File: logger/include/logger.hpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                       << " Line no: " << location.line() << " ";
+        }
+        return *this;
+    }
+    /**
+     * @brief Operator overloading function to append message to the buffer
+     *
+     * param[in]- Information to be appended to the buffer
+     */
+
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:47]: No empty lines between doxygen and code.

File: logger/include/logger.hpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            logmessage << "File: "
+                       << std::filesystem::path(location.file_name())
+                              .filename()
+                              .string()
+                       << " Function name: " << location.function_name()
+                       << " Line no: " << location.line() << " ";
+        }
+        return *this;
+    }
+    /**
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:42]: A blank line between blocks of code.

File: logger/include/logger.hpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    CRITICAL
+};
+
+class Logger
+{
+  public:
+    Loglevel level;
+    std::stringstream logmessage;
+    /**
+     * @brief checks the loglevel of the message
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:20]: Wrap nearest to 80 characters and please follow coding guidelines.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
We will need you to walk through the code for review.

File: logger/include/logger.hpp (hunk 5 above)
Comment (RekhaAparna01) [reply logger/include/logger.hpp:47]: The data type of message can be int, char, or string. so to keep it generic I used template

File: logger/include/logger.hpp (hunk 4 above)
Comment (RekhaAparna01) [reply logger/include/logger.hpp:48]: The data type of message can be string, char or int. so to keep it generic I used template.

File: logger/include/logger.hpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+#include <iostream>
+#include <source_location>
+
+namespace logging
+{
+
+/**
+ *@brief An API to append log messages to a buffer
+ *This API is called to log message. It also appends the
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp]: Add a blank line here.

Comment (SunnySrivastava1984) [reply logger/include/logger.hpp]: Also please re-phrase the description.

File: logger/include/logger.hpp [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ *file name and line number of the mesaage.
+ *
+ *param[in] - Information that we want to log
+ *param[in] - Object of source_location class
+ */
+void logMessage(std::string message, const std::source_location& location =
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp]: Please use constant reference as the string is not intended to be modified.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
I would like to see how the message format looks here.
Please add output of your code in the tested section of commit message. Also please refer the coding guidelines document and adhere to the guidelines for commit message.

File: logger/include/logger.hpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    if (Loglevel::INFO == level)
+    {
+        buffer.push_back(
+            " line no: " + std::to_string(location.line()) + " File: " +
+            std::filesystem::path(location.file_name()).filename().string() +
+            " Function Name: " + location.function_name() + " " + message +
+            "\n");
+    }
+
+    while (buffer.size() > size)
```
Comment (PriyangaRamasamy) [inline logger/include/logger.hpp:45]: why while() loop is required here? why can't just be an if() condition. 
I think the above buffer.push_back() pushes one line at a time, is that right?

File: logger/include/logger.hpp [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        buffer.erase(buffer.begin());
+    }
+}
+
+void logmessage(std::string message, const std::source_location& location =
```
Comment (PriyangaRamasamy) [inline logger/include/logger.hpp:51]: can you get message as const std::string&

File: logger/include/logger.hpp [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+                                         std::source_location::current())
+{
+    logmessage(Loglevel::INFO, message, location);
```
Comment (PriyangaRamasamy) [inline logger/include/logger.hpp:54]: am not following why we send this Loglevel::INFO

File: meson.build [hunk 13]
```diff
@@ -61,6 +61,11 @@ panel_tool_a = static_library(
      'tools/src/dbus_call.cpp',
      include_directories: ['include', 'tools/include']
 )
+logger_a = static_library(
```
Comment (PriyangaRamasamy) [inline meson.build:64]: why _a at the end of variable name?

**Review by PriyangaRamasamy — COMMENTED** [review]
Can you remove all the empty files

File: logger/include/logger.hpp [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+{
+    INFO,
+    CRITICAL
+};
+
+namespace Logger
+{
+
+std::list<std::string> buffer;
+const size_t size = 8000;
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply logger/include/logger.hpp:19]: why specific 8k size? need some detail here.

File: logger/include/logger.hpp [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+ *This API is called to log message to the buffer.
+ *It checks the log level and appends information like
+ *line number, file name, function name to the buffer.
+ *
+ *@param[in] - loglevel of the message
+ *@param[ni] - Information to be logged
+ *@param[in] - object of the source_location class
+ */
+void logmessage(
+    Loglevel level, const std::string message,
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply logger/include/logger.hpp:33]: if any of the input parameter is READ-only (means it doesn't need to modify its data in this function) , then keep it as "const"

File: logger/include/logger.hpp [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    const std::source_location& location = std::source_location::current())
+{
+    if (Loglevel::INFO == level)
```
[the remark this answers is not in the document]
Comment (Alpana07) [reply logger/include/logger.hpp:36]: are we handling only INFO level?

File: logger/include/logger.hpp (hunk 11 above)
Comment (Alpana07) [inline logger/include/logger.hpp:51]: logmessage-> logMessage

File: logger/include/logger.hpp [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ *This API is called to log message to the buffer.
+ *It checks the log level and appends information like
+ *line number, file name, function name to the buffer.
+ *
+ *@param[in] - loglevel of the message
+ *@param[ni] - Information to be logged
+ *@param[in] - object of the source_location class
+ */
+void logmessage(
```
Comment (Alpana07) [inline logger/include/logger.hpp:32]: logmessage-> logMessage

File: logger/include/logger.hpp [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    while (buffer.size() > size)
+    {
+        buffer.erase(buffer.begin());
+    }
+}
+
+void logmessage(std::string message, const std::source_location& location =
+                                         std::source_location::current())
```
Comment (Alpana07) [inline logger/include/logger.hpp:52]: I am not sure if we need something specific for level:: INFO?
If not, we don't need this wrapper call function . 
logmessage() can be called for all the levels. and if needed level wise handling also can be done there itself .

File: meson.build (hunk 13 above)
Comment (Alpana07) [reply meson.build:64]: > why _a at the end of variable name?

I think, she is following the existing pattern , as
panel_app_a = static_library(
panel_tool_a = static_library(

**Review by Alpana07 — COMMENTED** [review]
I have few comments ,please check

File: logger/include/logger.hpp (hunk 17 above)
Comment (RekhaAparna01) [reply logger/include/logger.hpp]: done

File: logger/include/logger.hpp (hunk 18 above)
Comment (RekhaAparna01) [reply logger/include/logger.hpp]: This is not specific for INFO level. If the log level is not specified while calling API, it takes INFO as the log level.

File: logger/include/logger.hpp (hunk 11 above)
Comment (RekhaAparna01) [reply logger/include/logger.hpp]: done

File: logger/include/logger.hpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (RekhaAparna01) [reply logger/include/logger.hpp]: done

File: logger/include/logger.hpp [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ *@brief Appends logmessage to the buffer
+ *
+ *This API is called to log message to the buffer.
+ *It checks the log level and appends information like
+ *line number, file name, function name to the buffer.
+ *
+ *@param[in] - loglevel of the message
+ *@param[ni] - Information to be logged
```
[the remark this answers is not in the document]
Comment (RekhaAparna01) [reply logger/include/logger.hpp]: done

File: logger/include/logger.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (RekhaAparna01) [reply logger/include/logger.hpp:38]: Yes, As of now only INFO level messages are appended to the buffer. can change it later as required.

File: logger/include/logger.hpp [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+ *@param[in] location - object of the source_location class
+ */
+void logMessage(
+    const Loglevel level, const std::string& message,
+    const std::source_location& location = std::source_location::current())
+{
+    if (Loglevel::INFO == level)
+    {
+        buffer.push_back(
+            " line no: " + std::to_string(location.line()) + " File: " +
```
Comment (branupama) [inline logger/include/logger.hpp:41]: we need to use time stamp as well right ?

log with time-stamp looks like: 
2017-06-06:17:07:02,158 DEBUG    [log.py:11] This is a debug log
2023-11-10 15:30:45 [INFO] my_module: example.py:42 - This is an informational message.
2023-11-10 12:00:00 [INFO] [example.py:23] - Program started.

**Review by branupama — COMMENTED** [review]
can we keep declarations in hpp file and definition in cpp file ?

File: logger/include/logger.hpp [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+{
+    INFO,
+    CRITICAL
+};
+
+namespace Logger
+{
+
+// size of buffer is 8k
+const size_t bufferSize = 8000;
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:20]: maxBufferSize is more readable.

File: logger/include/logger.hpp [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+inline std::list<std::string> buffer;
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:21]: Better name for buffer.

File: logger/include/logger.hpp [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+static std::map<Loglevel, std::string> log = {{Loglevel::INFO, "INFO"},
+                                              {Loglevel::CRITICAL, "CRITICAL"}};
+/**
+ *@brief Appends data to the buffer
+ *
+ *This API is called to log message to the buffer.
+ *It checks the log level and appends information like
+ *line number, file name, timestamp to the buffer.
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:29]: All these data are getting appended to the message, not to the buffer. Please re-phrase.

File: logger/include/logger.hpp [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ *@param[in] level - loglevel of the message
+ *@param[in] message - Information to be logged
+ *@param[in] location - object of the source_location class
+ */
+void logMessage(
+    const Loglevel level, const std::string& message,
+    const std::source_location& location = std::source_location::current());
+
+void logMessage(
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:39]: Where is the doxygen for this API?

File: logger/include/logger.hpp [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+    CRITICAL
+};
+
+namespace Logger
+{
+
+// size of buffer is 8k
+const size_t bufferSize = 8000;
+inline std::list<std::string> buffer;
+static std::map<Loglevel, std::string> log = {{Loglevel::INFO, "INFO"},
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:22]: Why is this required here? I don't see it being used in header file anywhere.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
Along with other comments,
A "why" should be added to commit message, explaining in brief why this change/implementation is required.

File: logger/include/logger.hpp [hunk 26]
```diff
@@ -1,20 +1,49 @@
...
+ */
+void logMessage(
+    const Loglevel level, const std::string& message,
+    const std::source_location& location = std::source_location::current());
+
 /**
- *@brief An API to append log messages to a buffer
- *This API is called to log message. It also appends the
- *file name and line number of the mesaage.
+ *@brief calls the logMessage(level, message, location) with INFO as default
```
Comment (SunnySrivastava1984) [inline logger/include/logger.hpp:39]: Please wrap brief in one line.
The API is to log message. What it calls internally need not be defined in brief. Though you can add some details after brief

Eg: An API to log message with log level defaulted to info.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: logger/src/logger.cpp [hunk 27]
```diff
@@ -1,18 +1,59 @@
...
+    }
+}
+
+void logMessage(const std::string& message,
+                const std::source_location& location)
 {
-    logMsg << "File: "
-           << std::filesystem::path(location.file_name()).filename().string()
-           << " line: " << location.line() << " " << message << "\n";
+    logMessage(Loglevel::INFO, message, location);
```
[the remark this answers is not in the document]
Comment (RekhaAparna01) [reply logger/src/logger.cpp:56]: Yes, It is required to have location object in this API. 

This api is going to call logMessage() internally with the message to be appended to the buffer. Here the message is coming from a different file.
so, if the location object is not passed the "location" parameter of logMessage() considers the "logger.cpp" as the file name instead of the original caller.

File: logger/include/logger.hpp (hunk 26 above)
Comment (RekhaAparna01) [reply logger/include/logger.hpp]: done

File: logger/src/logger.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (SunnySrivastava1984) [reply logger/src/logger.cpp:56]: Correct, we cal close this comment.

**Review by SunnySrivastava1984 — APPROVED** [review]

**Review by branupama — APPROVED** [review]

**Review by jinuthomas — APPROVED** [review]
